### PR TITLE
[inductor] Reindex transpose-contiguous clone to enable vertical fusion

### LINKED ISSUE #188635: torch.compile misses fusion for producer -> view/transpose/contiguous layout clone in attention-style paths
## Bug / Optimization Request

`torch.compile` with Inductor currently emits a separate clone/copy kernel for a common attention layout transition:

```python
embedding -> rmsnorm -> rope -> view(B, S, heads, D) -> transpose(1, 2).contiguous()
```

The producer writes a contiguous `[B, S, H]` buffer. The clone reads the same storage through the logical `[B, heads, S, D]` transpose order and writes a final contiguous output. The existing loop-index inversion path does not handle this case because the clone read/write dependency uses a 4D logical domain while the producer writes a 3D domain. As a result, Inductor fails vertical fusion and launches:

1. one fused producer kernel
2. one separate `triton_poi_fused_clone_*` copy kernel

## Reproducer

```python
import torch
import torch.nn.functional as F


def rotate_every_two(x):
    return torch.stack((-x[..., 1::2], x[..., 0::2]), dim=-1).flatten(-2)


def model(token_ids, embedding_weight, norm_weight, cos, sin):
    x = F.embedding(token_ids, embedding_weight)
    inv_rms = torch.rsqrt(torch.mean(x * x, dim=-1, keepdim=True) + 1e-6)
    x = x * inv_rms * norm_weight
    x = (x * cos) + (rotate_every_two(x) * sin)
    x = x.view(token_ids.shape[0], token_ids.shape[1], 4, 8)
    return x.transpose(1, 2).contiguous()


torch.manual_seed(0)
bsz, seq, vocab, hidden = 2, 16, 128, 32
token_ids = torch.randint(0, vocab, (bsz, seq), device="cuda", dtype=torch.long)
embedding_weight = torch.randn(vocab, hidden, device="cuda")
norm_weight = torch.randn(hidden, device="cuda")
pos = torch.arange(seq, device="cuda", dtype=torch.float32)[:, None]
dim = torch.arange(hidden, device="cuda", dtype=torch.float32)[None, :]
freqs = pos / (10000.0 ** (dim / hidden))
cos = torch.cos(freqs)[None, :, :]
sin = torch.sin(freqs)[None, :, :]

compiled = torch.compile(model, backend="inductor", fullgraph=True)
out = compiled(token_ids, embedding_weight, norm_weight, cos, sin)
torch.cuda.synchronize()
print(out.shape, out.stride(), float(out.sum()))
```

## Expected Behavior

Inductor should reindex the layout-changing clone to the producer iteration domain and fuse it vertically, so the generated code directly writes the final `[B, heads, S, D]` contiguous layout and does not launch a separate clone kernel.

## Observed Behavior

Without the optimization, generated code contains two launches:

```text
triton_per_fused_add_embedding_mean_mul_rsqrt_0.run(...)
triton_poi_fused_clone_1.run(...)
```

## Environment Used for Measurement

See `collect_env.txt` in the attached PR prep package.

Short version:
- PyTorch 2.5.1+cu121 installed package, Python-layer patched for experiment
- Triton 3.1.0
- NVIDIA GeForce RTX 4090
- CUDA build 12.1

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by timeuser4**
Fixes #188635

## Summary

This PR extends Inductor's loop reindexing/index-inversion fusion path to handle a layout-changing clone produced by:

```python
producer[B, S, H] -> view(B, S, heads, D) -> transpose(1, 2).contiguous()
```

Before this change, the producer writes `[B, S, H]` contiguously, while the clone reads the producer buffer through a 4D `[B, heads, S, D]` logical domain and writes the final contiguous output. Because the read/write dependency domains differ, vertical fusion rejects the producer -> clone edge and Inductor emits a separate `triton_poi_fused_clone_*` copy kernel.

This change adds:

- `LoopBody.reindex_iter_loops_with_indexer(...)`, a small extension that allows reindexing with an explicit new-index -> old-index mapping.
- `SchedulerNode.apply_loop_reindexing_with_indexer(...)`, matching the existing `apply_loop_reindexing(...)` lifecycle and dependency refresh behavior.
- A targeted scheduler helper for the `[B, S, H] -> [B, heads, S, D]` transpose-contiguous clone pattern. It reindexes the clone to the producer domain `[B, S, H]`, making the clone read dependency match the producer write while preserving the final output store indexing.
- A regression test covering `embedding -> rmsnorm -> rope -> view -> transpose -> contiguous`.

The optimization is guarded narrowly:

- CUDA/GPU path only through the existing caller.
- `node2` must be a non-reduction `SchedulerNode`.
- clone must have one read and one write.
- body must have exactly two indexing expressions and no subblocks.
- producer domain must be 3D and clone read/write domain must be matching 4D.
- sizes must satisfy `B == B`, `S == S`, and `H == heads * head_dim`.
- if the reindex does not improve fusion score, node state is restored.

## Test Plan

Main checkout patch/static checks, run in `/home/admin/app/pytorch-dev` at `703abdc`:

```bash
git apply --check pytorch_layout_clone_reindex_pr.patch
git diff --check
python3 -m py_compile torch/_inductor/loop_body.py torch/_inductor/scheduler.py test/inductor/test_loop_ordering.py
```

Target test from the main git checkout, run through a wrapper because the available runtime is an older `torch==2.5.1+cu121` wheel while the checkout is current main:

```bash
TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION=1 \
TORCHINDUCTOR_LOOP_INDEX_INVERSION_IN_FUSION=1 \
TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 \
TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 \
python3 /tmp/run_main_checkout_target_test.py
```

Result:

```text
Ran 1 test in 3.236s
OK
```

Runtime-compatible PyTorch source tree check, run from `/home/admin/app/pytorch-v2.5.1-src` using the installed 2.5.1 CUDA wheel's generated/binary files:

```bash
TORCHINDUCTOR_LOOP_REINDEXING_AFTER_FUSION=1 \
TORCHINDUCTOR_LOOP_INDEX_INVERSION_IN_FUSION=1 \
TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 \
TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 \
PYTHONPATH=/home/admin/app/pytorch-v2.5.1-src \
python3 test/inductor/test_loop_ordering.py -k test_transpose_contiguous_clone_reindexing
```

Result:

```text
Ran 1 test in 3.047s
OK
```

Expected validation before opening the final GitHub PR from a fully built PyTorch main dev environment:

```bash
python test/inductor/test_loop_ordering.py -k test_transpose_contiguous_clone_reindexing
python test/inductor/test_loop_ordering.py -k "reindexing or inversion"
python -m tools.linter.lintrunner --take CLANGFORMAT MYPYNOFOLLOW FLAKE8
```

Experiment run in `4090-2` / `puyun-acc` using the installed PyTorch package with the same Python-layer behavior:

Correctness:

```text
max_abs_diff 4.76837158203125e-07
out_sum 11.314574241638184
stride (512, 128, 8, 1)
```

Fusion evidence:

```text
transpose-contiguous clone reindex matched op1 -> op2 with score 4096
completed fusion round (1/10): fused 3 nodes into 1 nodes
```

Generated code evidence:

```text
triton_per_fused_add_clone_embedding_mean_mul_rsqrt_0.run(...)
```

No separate `triton_poi_fused_clone_*` launch remains.

Benchmark summary, float32, RTX 4090:

```text
small:  default torch.compile median 0.125744 ms; replaced median 0.109488 ms
medium: default torch.compile median 0.157328 ms; replaced median 0.094576 ms
```

Full logs are in:

- `pytorch_main_checkout_wrapped_target_test.log`
- `pytorch_dev_checkout_test_transpose_contiguous_clone_reindexing.log`
- `experiment_artifacts/probe_fusion_log.txt`
- `experiment_artifacts/output_code_replaced.py`
- `experiment_artifacts/ir_post_fusion_replaced.txt`
- `experiment_artifacts/benchmark_default_two_kernel.txt`
- `experiment_artifacts/benchmark_reindexed_replaced.txt`
- `experiment_artifacts/check_replaced_correctness.out`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by timeuser4** [comment]
Addressed Copilot's rollback concern by comparing the post-reindex fusion score against the baseline score and restoring the node state unless the score strictly improves.

Re-ran:
- `git diff --check`
- `python3 -m py_compile torch/_inductor/loop_body.py torch/_inductor/scheduler.py test/inductor/test_loop_ordering.py`
- target `test_transpose_contiguous_clone_reindexing` via the same main-checkout compatibility wrapper: `Ran 1 test ... OK`

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by timeuser4** [comment]
@jansel   CI is green. Could this get another look when convenient?
Please let me know if there is anything else I should address, or if there are concerns about the current approach.

**Comment by jansel** [comment]
@claude review these changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by timeuser4** [comment]
All requested changes are now addressed.

Regarding Claude’s concerns about dynamic B/S and the hardcoded index polynomials, I’ve kept the targeted matcher for the current attention layout because generalizing it to dynamic shapes and broader polynomial forms would expand the scope and require substantially broader tests. For the current pattern, statically_known_equals() already handles equivalent simplifications. I’ve added a note documenting this choice. If that works, I’ll open a follow-up PR to address broader matching. @jansel

**Comment by jansel** [comment]
@claude review these changes again

**Review by jansel — APPROVED** [review]

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -7103,6 +7117,119 @@ def shared_data_after_inverting_indexing(
...
+        if len(node2.read_writes.reads) != 1 or len(node2.read_writes.writes) != 1:
+            return -1
+
+        body = node2._body
+        if not body.is_memory_copy():
+            return -1
+        if len(body.indexing_exprs) != 2 or body.subblocks:
+            return -1
+        if "index0" not in body.indexing_exprs or "index1" not in body.indexing_exprs:
+            return -1
```
Comment (eellison) [inline torch/_inductor/scheduler.py]: I think "index0" and "index1" is a bit of a code smell. 

Also, memory copy doesn't preclude layout/indexing.

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -7103,6 +7117,119 @@ def shared_data_after_inverting_indexing(
...
+
+        producer_size = list(node1_write.size)
+        clone_read_size = list(node2_read.size)
+        clone_write_size = list(node2_write.size)
+        if (
+            len(producer_size) != 3
+            or len(clone_read_size) != 4
+            or clone_read_size != clone_write_size
+        ):
+            return -1
```
Comment (eellison) [inline torch/_inductor/scheduler.py]: We should not have to special case 3/4 in the reindexing  or index inversion.

**Review by eellison — COMMENTED** [review]
thanks for pr and issue, i think there are better ways of doing this that does not specifically encode size 3 -> size 4.

will get back with more substantive suggestions

**Review by eellison — CHANGES_REQUESTED** [review]
sorry for late review, should be more general, not just handle 3-> 4

**Comment by timeuser4** [comment]
@claude review these changes again

**Review by eellison — CHANGES_REQUESTED** [review]
oops, sorry, errant comment

**Review by eellison — DISMISSED** [review]
The generalized implementation addresses my original concern. I am okay with landing this direction and handling remaining integration follow-ups in the subsequent stack. The current CUDA/ROCm test failures still need to be resolved or otherwise cleared by CI before merge.

**Comment by eellison** [comment]
@timeuser4 https://github.com/pytorch/pytorch/pull/190402 fixes https://github.com/pytorch/pytorch/issues/188635. I had a use case already for a similar transformation and i verified it fixed your issue as well. 

Sorry for the churn here in landing. The scheduling, reindexing / inversion code is a bit tricky for a first time contributor. If you are interested in contributing or landing other prs let me know and I can suggest other issues.

**Comment by timeuser4** [comment]
@eellison  Thank you so much for your help. I understand the approach now. I was too focused on solving the local special case to avoid expanding the scope, rather than normalizing the consumer's iteration domain and reusing the existing index-inversion machinery. I really enjoyed working through this problem. Could you recommend any other Inductor issues? I'd be excited to work on more Inductor issues

**Comment by eellison** [comment]
@timeuser4 I would just pick up unassigned issues in inductor, and try working on them. see: https://github.com/pytorch/pytorch/issues?q=sort%3Aupdated-desc%20is%3Aissue%20state%3Aopen%20label%3A%22module%3A%20inductor%22%20-label%3A%22module%3A%20flaky-tests%22%20-label%3A%22module%3A%20tests%22%20-label%3A%22module%3A%20ci%22%20no%3Aassignee%20-label%3A%22module%3A%20cpu%22%20-label%3Alarge%20-label%3Askipped
